src/mame/sinclair/specnext.cpp

```diff
@@ -0,0 +1,2757 @@
+	m_regs.nr_92_pi_gpio_o_en = 0x00;
+	m_regs.nr_93_pi_gpio_o_en = 0x00;
+
+	m_regs.nr_a0_pi_peripheral_en = 0x00;
+	m_regs.nr_a2_pi_i2s_ctl = 0x00;
+
+	m_regs.nr_a8_esp_gpio0_en = 0;
+	m_regs.nr_a9_esp_gpio0 = 1;
+
+	m_regs.nr_b8_divmmc_ep_0 = 0x83;
+	m_regs.nr_b9_divmmc_ep_valid_0 = 0x01;
+	m_regs.nr_ba_divmmc_ep_timing_0 = 0x00;
+	m_regs.nr_bb_divmmc_ep_1 = 0xcd;
+
+	m_regs.nr_c0_im2_vector = 0x00;
+	m_regs.nr_c0_stackless_nmi = 0;
+	m_maincpu->nmi_stackless_w(m_regs.nr_c0_stackless_nmi);
+	m_regs.nr_c0_int_mode_pulse_0_im2_1 = 0;
+
+	m_regs.nr_c4_int_en_0_expbus = 1;
+
+	m_regs.nr_c6_int_en_2_654 = 0x00;
+	m_regs.nr_c6_int_en_2_210 = 0x00;
+
+	m_regs.nr_cc_dma_int_en_0_7 = 0;
+	m_regs.nr_cc_dma_int_en_0_10 = 0x00;
+	m_regs.nr_cd_dma_int_en_1 = 0x00;
+	m_regs.nr_ce_dma_int_en_2_654 = 0x00;
+	m_regs.nr_ce_dma_int_en_2_210 = 0x00;
+
+	m_regs.nr_d8_io_trap_fdc_en = 0;
+
+	if (m_regs.nr_03_config_mode)
+		m_bootrom_en = 1;
+
+	mmu_w(0, 0xff);
+	mmu_w(1, 0xff);
+	mmu_w(2, 0x0a);
+	mmu_w(3, 0x0b);
+	mmu_w(4, 0x04);
+	mmu_w(5, 0x05);
+	mmu_w(6, 0x00);
+	mmu_w(7, 0x01);
+}
+
+static const gfx_layout bootrom_charlayout =
+{
+	8, 8,          // 8 x 8 characters */
+	96,            // 96 characters */
+	1,             // 1 bits per pixel */
+	{ 0 },           // no bitplanes
+	{ STEP8(0, 1) }, // x offsets
+	{ STEP8(0, 8) }, // y offsets
+	8 * 8          // every char takes 8 bytes
+};
+
+static GFXDECODE_START(gfx_tbblue)
+	GFXDECODE_ENTRY("maincpu", 0x1584, bootrom_charlayout, 0xf7, 1)
+GFXDECODE_END
+
+void specnext_state::video_start()
+{
+	spectrum_128_state::video_start();
+
+	address_space &prg = m_maincpu->space(AS_PROGRAM);
+	// see (1)
+	prg.install_write_tap(0x0000, 0xbfff, "layer2_w", [this](offs_t offset, u8 &data, u8 mem_mask)
+	{
+		if (m_regs.port_123b_layer2_map_wr_en)
+		{
+			const u8 type = m_regs.port_123b_layer2_map_segment;
+			u8* to = m_ram->pointer() + 0x40000 + (layer2_active_bank() << 14);
+			if ((type < 3) && (offset < 0x4000))
+				to[offset + (type << 14)] = data;
+			else // if (type == 3)
+				to[offset] = data;
+		}
+	});
+	prg.install_write_tap(0x0000, 0x3fff, "altrom_w", [this](offs_t offset, u8 &data, u8 mem_mask)
+	{
+		if (nr_8c_altrom_en() && nr_8c_altrom_rw())
+		{
+			u8* altrom = m_ram->pointer() + ((0b00001100 | (m_sram_alt_128_n << 1)) << 13);
+			altrom[offset] = data;
+		}
+	});
+}
+
+void specnext_state::tbblue(machine_config &config)
+{
+	spectrum_128(config);
+	config.device_remove("exp");
+	// m_ram->set_default_size("1M").set_extra_options("2M");
+	m_ram->set_default_size("2M");
+
+	Z80N(config.replace(), m_maincpu, X1_128_SINCLAIR / 10);
+	m_maincpu->set_m1_map(&specnext_state::map_fetch);
+	m_maincpu->set_memory_map(&specnext_state::map_mem);
+	m_maincpu->set_io_map(&specnext_state::map_io);
+	m_maincpu->set_vblank_int("screen", FUNC(specnext_state::spec_interrupt)); // ...
+	m_maincpu->out_nextreg_cb().set(FUNC(specnext_state::reg_w));
+	m_maincpu->in_nextreg_cb().set(FUNC(specnext_state::reg_r));
```

Comment:
Would it make more sense to have an additional address CPU address space for these, and use individual handler functions?  The huge switch statements in functions aren’t great form.

Reply:
I found it's easier to modify such form. Address space will require function for each entry which makes code much more verbose. Currently I'm using lambdas if "function" is small enough but don't really like this way because the code looks inconsistent.

Reply:
![Screenshot_2024-01-09-02-37-45-71_3aea4af51f236e4932235fdada7d1643~2](https://github.com/mamedev/mame/assets/58155/4c1018a6-2c66-43d8-bf1b-152a0064e2eb)

Plus FPGA implements this as a gigantic case stetment, hance it's easier to track any further updates.